File: apps/guzo-api/src/main/java/et/guzo/service/DriverOpsService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.Driver;
import et.guzo.domain.entity.Order;
import et.guzo.domain.entity.TransportManifest;
import et.guzo.domain.entity.Vehicle;
import et.guzo.domain.entity.VehicleLog;
import et.guzo.domain.enums.ManifestStatus;
import et.guzo.domain.enums.OrderStatus;
import et.guzo.domain.enums.VehicleLogType;
import et.guzo.domain.enums.WalletTxnType;
import et.guzo.repository.DriverRepository;
import et.guzo.repository.ManifestParcelRepository;
import et.guzo.repository.TransportManifestRepository;
import et.guzo.repository.WalletTransactionRepository;
import et.guzo.repository.VehicleLogRepository;
import et.guzo.repository.VehicleRepository;
import et.guzo.repository.DeliveryRepository;
import et.guzo.repository.OrderRepository;
import et.guzo.repository.AddressRepository;
import et.guzo.domain.entity.Address;
import et.guzo.domain.entity.Delivery;
import et.guzo.util.IdUtil;
import et.guzo.security.AuthUser;
import et.guzo.security.RoleChecker;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.*;

@Service
@RequiredArgsConstructor
public class DriverOpsService {

    private final DriverRepository driverRepository;
    private final TransportManifestRepository manifestRepository;
    private final ManifestParcelRepository manifestParcelRepository;
    private final TransportManifestService manifestService;
    private final WalletTransactionRepository walletTransactionRepository;
    private final VehicleRepository vehicleRepository;
    private final VehicleLogRepository vehicleLogRepository;
    private final DeliveryRepository deliveryRepository;
    private final OrderRepository orderRepository;
    private final AddressRepository addressRepository;

    private Driver requireDriver(String userId) {
        return driverRepository.findByUserId(userId)
            .orElseThrow(() -> ApiException.badRequest("Authenticated user is not a driver"));
    }

    private void assertManifestDriver(TransportManifest manifest, String driverId) {
        if (manifest.getDriverId() == null || !manifest.getDriverId().equals(driverId)) {
            throw ApiException.forbidden("This manifest is not assigned to you");
        }
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listManifests(String userId) {
        Driver driver = requireDriver(userId);
        List<ManifestStatus> statuses = List.of(ManifestStatus.DRAFT, ManifestStatus.SEALED, ManifestStatus.IN_TRANSIT, ManifestStatus.ARRIVED);
        return manifestRepository.findByDriverIdAndStatusInOrderByCreatedAtDesc(driver.getId(), statuses).stream()
            .map(m -> {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("id", m.getId());
                row.put("manifestNumber", m.getManifestNumber());
                row.put("status", m.getStatus().name());
                row.put("sealNumber", m.getSealNumber());
                row.put("parcelCount", manifestParcelRepository.findByManifestId(m.getId()).size());
                row.put("departedAt", m.getDepartedAt());
                row.put("arrivedAt", m.getArrivedAt());
                return row;
            }).toList();
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getManifest(String userId, String manifestId) {
        Driver driver = requireDriver(userId);
        TransportManifest manifest = manifestRepository.findById(manifestId)
            .orElseThrow(() -> ApiException.notFound("Manifest not found"));
        assertManifestDriver(manifest, driver.getId());
        return manifestService.getDetail(manifestId);
    }

    @Transactional
    public Object scanManifest(String userId, String manifestId, String trackingNumber) {
        Driver driver = requireDriver(userId);
        TransportManifest manifest = manifestRepository.findById(manifestId)
            .orElseThrow(() -> ApiException.notFound("Manifest not found"));
        assertManifestDriver(manifest, driver.getId());
        return manifestService.scanParcel(manifestId, null, trackingNumber, userId);
    }

    @Transactional
    public TransportManifest departManifest(String userId, String manifestId, String sealNumber) {
        Driver driver = requireDriver(userId);
        TransportManifest manifest = manifestRepository.findById(manifestId)
            .orElseThrow(() -> ApiException.notFound("Manifest not found"));
        assertManifestDriver(manifest, driver.getId());
        return manifestService.sealAndDepart(manifestId, sealNumber != null ? sealNumber : "SEAL-" + System.currentTimeMillis());
    }

    @Transactional
    public TransportManifest arriveManifest(String userId, String manifestId) {
        Driver driver = requireDriver(userId);
        TransportManifest manifest = manifestRepository.findById(manifestId)
            .orElseThrow(() -> ApiException.notFound("Manifest not found"));
        assertManifestDriver(manifest, driver.getId());
        return manifestService.markArrived(manifestId);
    }

    @Transactional
    public Map<String, Object> unloadManifest(AuthUser user, String manifestId, String trackingNumber) {
        Driver driver = requireDriver(user.getId());
        TransportManifest manifest = manifestRepository.findById(manifestId)
            .orElseThrow(() -> ApiException.notFound("Manifest not found"));
        assertManifestDriver(manifest, driver.getId());
        return manifestService.unloadScan(manifestId, trackingNumber, user);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getEarnings(String userId) {
        Driver driver = requireDriver(userId);
        var txns = walletTransactionRepository.findByUserIdOrderByCreatedAtDesc(driver.getUserId(), PageRequest.of(0, 50));
        List<Map<String, Object>> transactions = txns.getContent().stream()
            .filter(t -> t.getType() == WalletTxnType.CREDIT)
            .map(t -> Map.<String, Object>of(
                "id", t.getId(),
                "amount", t.getAmount(),
                "balanceAfter", t.getBalanceAfter(),
                "currency", t.getCurrency(),
                "reference", t.getReference() != null ? t.getReference() : "",
                "description", t.getDescription() != null ? t.getDescription() : "",
                "createdAt", t.getCreatedAt().toString()
            )).toList();
        return Map.of(
            "balance", driver.getEarningsBalance(),
            "totalDeliveries", driver.getTotalDeliveries(),
            "transactions", transactions
        );
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getMyVehicle(String userId) {
        Driver driver = requireDriver(userId);
        return vehicleRepository.findFirstByDriverIdOrderByCreatedAtDesc(driver.getId())
            .map(v -> {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("id", v.getId());
                row.put("plateNumber", v.getPlateNumber());
                row.put("type", v.getType().name());
                row.put("status", v.getStatus().name());
                row.put("brand", v.getBrand());
                row.put("model", v.getModel());
                row.put("color", v.getColor());
                return row;
            })
            .orElse(null);
    }

    @Transactional
    public Map<String, Object> createVehicleLog(String userId, VehicleLogType type, BigDecimal odometerKm, BigDecimal amount, String note) {
        Driver driver = requireDriver(userId);
        Vehicle vehicle = vehicleRepository.findFirstByDriverIdOrderByCreatedAtDesc(driver.getId())
            .orElseThrow(() -> ApiException.badRequest("No vehicle assigned to your driver profile"));
        Instant now = Instant.now();
        VehicleLog log = new VehicleLog();
        log.setId(IdUtil.cuid());
        log.setVehicleId(vehicle.getId());
        log.setDriverId(driver.getId());
        log.setType(type);
        log.setOdometerKm(odometerKm);
        log.setAmount(amount);
        log.setNote(note);
        log.setLoggedAt(now);
        log.setCreatedAt(now);
        vehicleLogRepository.save(log);
        return Map.of(
            "id", log.getId(),
            "type", log.getType().name(),
            "odometerKm", log.getOdometerKm(),
            "amount", log.getAmount(),
            "note", log.getNote() != null ? log.getNote() : "",
            "loggedAt", log.getLoggedAt().toString()
        );
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listVehicleLogs(String userId, int limit) {
        Driver driver = requireDriver(userId);
        return vehicleLogRepository.findByDriverIdOrderByLoggedAtDesc(driver.getId(), PageRequest.of(0, limit)).stream()
            .map(l -> {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("id", l.getId());
                row.put("type", l.getType().name());
                row.put("odometerKm", l.getOdometerKm());
                row.put("amount", l.getAmount());
                row.put("note", l.getNote());
                row.put("loggedAt", l.getLoggedAt().toString());
                return row;
            }).toList();
    }

    @Transactional(readOnly = true)
    public Map<String, Object> optimizedRoute(String userId) {
        Driver driver = requireDriver(userId);
        List<Delivery> deliveries = deliveryRepository.findByDriverIdAndDeliveredAtIsNull(driver.getId());

        Set<OrderStatus> pickupStatuses = EnumSet.of(OrderStatus.ASSIGNED, OrderStatus.CONFIRMED, OrderStatus.PICKED_UP);
        Set<OrderStatus> dropStatuses = EnumSet.of(OrderStatus.OUT_FOR_DELIVERY, OrderStatus.IN_TRANSIT);

        List<Map<String, Object>> candidates = new ArrayList<>();
        for (Delivery d : deliveries) {
            Order order = orderRepository.findById(d.getOrderId()).orElse(null);
            if (order == null) continue;
            if (pickupStatuses.contains(order.getStatus())) {
                Address addr = addressRepository.findById(order.getPickupAddressId()).orElse(null);
                if (addr != null) {
                    candidates.add(stop(order, "pickup", addr));
                }
            }
            if (dropStatuses.contains(order.getStatus())) {
                Address addr = addressRepository.findById(order.getDropoffAddressId()).orElse(null);
                if (addr != null) {
                    candidates.add(stop(order, "dropoff", addr));
                }
            }
        }

        List<Map<String, Object>> withCoords = new ArrayList<>();
        List<Map<String, Object>> withoutCoords = new ArrayList<>();
        for (Map<String, Object> c : candidates) {
            if (c.get("latitude") != null && c.get("longitude") != null) {
                withCoords.add(c);
            } else {
                withoutCoords.add(c);
            }
        }

        Double curLat = driver.getCurrentLat();
        Double curLng = driver.getCurrentLng();
        if ((curLat == null || curLng == null) && !withCoords.isEmpty()) {
            curLat = (Double) withCoords.get(0).get("latitude");
            curLng = (Double) withCoords.get(0).get("longitude");
        }

        List<Map<String, Object>> ordered = new ArrayList<>();
        List<Map<String, Object>> remaining = new ArrayList<>(withCoords);
        double totalKm = 0;

        while (!remaining.isEmpty() && curLat != null && curLng != null) {
            int bestIdx = 0;
            double bestDist = Double.MAX_VALUE;
            for (int i = 0; i < remaining.size(); i++) {
                Map<String, Object> s = remaining.get(i);
                double dist = haversineKm(curLat, curLng, (Double) s.get("latitude"), (Double) s.get("longitude"));
                if (dist < bestDist) {
                    bestDist = dist;
                    bestIdx = i;
                }
            }
            Map<String, Object> next = remaining.remove(bestIdx);
            totalKm += bestDist == Double.MAX_VALUE ? 0 : bestDist;
            ordered.add(next);
            curLat = (Double) next.get("latitude");
            curLng = (Double) next.get("longitude");
        }

        List<Map<String, Object>> stops = new ArrayList<>(ordered);
        stops.addAll(withoutCoords);

        return Map.of(
            "stops", stops,
            "totalStops", stops.size(),
            "estimatedKm", Math.round(totalKm * 100.0) / 100.0
        );
    }

    private static Map<String, Object> stop(Order order, String type, Address addr) {
        Map<String, Object> row = new LinkedHashMap<>();
        row.put("orderId", order.getId());
        row.put("orderNumber", order.getOrderNumber());
        row.put("type", type);
        row.put("line1", addr.getLine1());
        row.put("city", addr.getCity());
        row.put("latitude", addr.getLatitude());
        row.put("longitude", addr.getLongitude());
        return row;
    }

    private static double haversineKm(double lat1, double lon1, double lat2, double lon2) {
        double r = 6371;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
            + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
        return r * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    }
}

File: apps/server/src/modules/drivers/driver-ops.service.ts
import { prisma } from '@delivery/database';
import { haversineKm } from '@delivery/utils';
import { ApiError } from '../../utils/ApiError.js';
import { manifestsService } from '../manifests/manifests.service.js';
import { storage } from '../../providers/storage/index.js';
import { UPLOAD_FOLDERS } from '../../constants/index.js';
import { filePublicUrl } from '../../utils/file-url.js';

export class DriverOpsService {
  private async requireDriver(userId: string) {
    const driver = await prisma.driver.findUnique({ where: { userId } });
    if (!driver) throw ApiError.badRequest('Authenticated user is not a driver');
    return driver;
  }

  private async assertManifestDriver(manifestId: string, driverId: string) {
    const manifest = await prisma.transportManifest.findUnique({ where: { id: manifestId } });
    if (!manifest) throw ApiError.notFound('Manifest not found');
    if (manifest.driverId !== driverId) {
      throw ApiError.forbidden('This manifest is not assigned to you');
    }
    return manifest;
  }

  async listManifests(userId: string) {
    const driver = await this.requireDriver(userId);
    const rows = await prisma.transportManifest.findMany({
      where: { driverId: driver.id, status: { in: ['DRAFT', 'SEALED', 'IN_TRANSIT', 'ARRIVED'] } },
      orderBy: { createdAt: 'desc' },
      include: { _count: { select: { parcels: true } } },
    });
    return rows.map((m) => ({
      id: m.id,
      manifestNumber: m.manifestNumber,
      status: m.status,
      sealNumber: m.sealNumber,
      originWarehouseId: m.originWarehouseId,
      destinationWarehouseId: m.destinationWarehouseId,
      parcelCount: m._count.parcels,
      departedAt: m.departedAt,
      arrivedAt: m.arrivedAt,
    }));
  }

  async getManifest(userId: string, manifestId: string) {
    const driver = await this.requireDriver(userId);
    await this.assertManifestDriver(manifestId, driver.id);
    return manifestsService.getDetail(manifestId);
  }

  async scanManifest(userId: string, manifestId: string, trackingNumber: string) {
    const driver = await this.requireDriver(userId);
    await this.assertManifestDriver(manifestId, driver.id);
    return manifestsService.scanParcel(manifestId, { trackingNumber, scannedByUserId: userId });
  }

  async departManifest(userId: string, manifestId: string, sealNumber?: string) {
    const driver = await this.requireDriver(userId);
    await this.assertManifestDriver(manifestId, driver.id);
    return manifestsService.depart(manifestId, sealNumber ?? `SEAL-${Date.now()}`);
  }

  async arriveManifest(userId: string, manifestId: string) {
    const driver = await this.requireDriver(userId);
    await this.assertManifestDriver(manifestId, driver.id);
    return manifestsService.arrive(manifestId);
  }

  async unloadManifest(userId: string, manifestId: string, trackingNumber: string) {
    const driver = await this.requireDriver(userId);
    await this.assertManifestDriver(manifestId, driver.id);
    return manifestsService.unloadScan(manifestId, trackingNumber);
  }

    async getEarnings(userId: string) {
    const driver = await this.requireDriver(userId);
    const transactions = await prisma.walletTransaction.findMany({
      where: { userId: driver.userId, type: 'CREDIT' },
      orderBy: { createdAt: 'desc' },
      take: 50,
    });
    return {
      balance: Number(driver.earningsBalance),
      totalDeliveries: driver.totalDeliveries,
      transactions: transactions.map((t) => ({
        id: t.id,
        amount: Number(t.amount),
        balanceAfter: Number(t.balanceAfter),
        currency: t.currency,
        reference: t.reference,
        description: t.description,
        createdAt: t.createdAt.toISOString(),
      })),
    };
  }

    async getMyVehicle(userId: string) {
    const driver = await this.requireDriver(userId);
    const vehicle = await prisma.vehicle.findFirst({
      where: { driverId: driver.id },
      orderBy: { updatedAt: 'desc' },
    });
    if (!vehicle) return null;
    const photoFile = vehicle.photoFileId
      ? await prisma.file.findUnique({ where: { id: vehicle.photoFileId } })
      : null;
    return {
      id: vehicle.id,
      plateNumber: vehicle.plateNumber,
      type: vehicle.type,
      status: vehicle.status,
      brand: vehicle.brand,
      model: vehicle.model,
      color: vehicle.color,
      photoUrl: filePublicUrl(photoFile?.storageKey),
    };
  }

  async upsertMyVehicle(
    userId: string,
    input: {
      type: string;
      plateNumber: string;
      brand?: string;
      model?: string;
      color?: string;
    },
  ) {
    const driver = await this.requireDriver(userId);
    const plate = input.plateNumber.trim().toUpperCase();
    if (!plate) throw ApiError.badRequest('Plate number is required');

    const existing = await prisma.vehicle.findFirst({ where: { driverId: driver.id } });
    const data = {
      type: input.type as never,
      plateNumber: plate,
      brand: input.brand?.trim() || null,
      model: input.model?.trim() || null,
      color: input.color?.trim() || null,
      status: 'ACTIVE' as const,
    };

    const vehicle = existing
      ? await prisma.vehicle.update({ where: { id: existing.id }, data })
      : await prisma.vehicle.create({ data: { ...data, driverId: driver.id } });

    const photoFile = vehicle.photoFileId
      ? await prisma.file.findUnique({ where: { id: vehicle.photoFileId } })
      : null;

    return {
      id: vehicle.id,
      plateNumber: vehicle.plateNumber,
      type: vehicle.type,
      status: vehicle.status,
      brand: vehicle.brand,
      model: vehicle.model,
      color: vehicle.color,
      photoUrl: filePublicUrl(photoFile?.storageKey),
    };
  }

  async uploadVehiclePhoto(
    userId: string,
    file: { path: string; filename: string; originalname: string; mimetype: string; size: number },
  ) {
    const driver = await this.requireDriver(userId);
    let vehicle = await prisma.vehicle.findFirst({ where: { driverId: driver.id } });
    if (!vehicle) {
      vehicle = await prisma.vehicle.create({
        data: {
          driverId: driver.id,
          type: 'MOTORCYCLE',
          plateNumber: `TMP-${driver.driverCode}`,
          status: 'ACTIVE',
        },
      });
    }

    const saved = await storage.save({
      absolutePath: file.path,
      folder: UPLOAD_FOLDERS.VEHICLE_PHOTOS,
      filename: file.filename,
    });

    const fileRow = await prisma.file.create({
      data: {
        uploaderId: userId,
        category: 'IMAGE',
        originalName: file.originalname,
        storedName: file.filename,
        mimeType: file.mimetype,
        sizeBytes: file.size,
        storageKey: saved.storageKey,
        storageDriver: saved.driver,
      },
    });

    const updated = await prisma.vehicle.update({
      where: { id: vehicle.id },
      data: { photoFileId: fileRow.id },
    });

    return {
      id: updated.id,
      plateNumber: updated.plateNumber,
      type: updated.type,
      status: updated.status,
      brand: updated.brand,
      model: updated.model,
      color: updated.color,
      photoUrl: filePublicUrl(saved.storageKey),
    };
  }

    async createVehicleLog(
    userId: string,
    input: {
      type: 'FUEL' | 'MAINTENANCE' | 'MILEAGE' | 'INSPECTION';
      odometerKm?: number;
      amount?: number;
      note?: string;
      metadata?: Record<string, unknown>;
    },
  ) {
    const driver = await this.requireDriver(userId);
    const vehicle = await prisma.vehicle.findFirst({ where: { driverId: driver.id } });
    if (!vehicle) throw ApiError.badRequest('No vehicle assigned to your driver profile');

    const log = await prisma.vehicleLog.create({
      data: {
        vehicleId: vehicle.id,
        driverId: driver.id,
        type: input.type,
        odometerKm: input.odometerKm,
        amount: input.amount,
        note: input.note,
        metadata: input.metadata as object | undefined,
      },
    });
    return {
      id: log.id,
      type: log.type,
      odometerKm: log.odometerKm != null ? Number(log.odometerKm) : null,
      amount: log.amount != null ? Number(log.amount) : null,
      note: log.note,
      loggedAt: log.loggedAt.toISOString(),
    };
  }

  async listVehicleLogs(userId: string, limit = 30) {
    const driver = await this.requireDriver(userId);
    const logs = await prisma.vehicleLog.findMany({
      where: { driverId: driver.id },
      orderBy: { loggedAt: 'desc' },
      take: limit,
    });
    return logs.map((l) => ({
      id: l.id,
      type: l.type,
      odometerKm: l.odometerKm != null ? Number(l.odometerKm) : null,
      amount: l.amount != null ? Number(l.amount) : null,
      note: l.note,
      loggedAt: l.loggedAt.toISOString(),
    }));
  }

    async optimizedRoute(userId: string) {
    const driver = await this.requireDriver(userId);
    const deliveries = await prisma.delivery.findMany({
      where: { driverId: driver.id, deliveredAt: null },
      include: {
        order: {
          include: {
            pickupAddress: true,
            dropoffAddress: true,
          },
        },
      },
    });

    type Stop = {
      orderId: string;
      orderNumber: string;
      type: 'pickup' | 'dropoff';
      line1: string;
      city: string;
      latitude: number | null;
      longitude: number | null;
    };

    const candidates: Stop[] = [];
    for (const d of deliveries) {
      const o = d.order;
      if (!o) continue;
      const pickupStatuses = ['ASSIGNED', 'CONFIRMED', 'PICKED_UP'];
      const dropStatuses = ['OUT_FOR_DELIVERY', 'IN_TRANSIT'];
      if (pickupStatuses.includes(o.status)) {
        candidates.push({
          orderId: o.id,
          orderNumber: o.orderNumber,
          type: 'pickup',
          line1: o.pickupAddress.line1,
          city: o.pickupAddress.city,
          latitude: o.pickupAddress.latitude,
          longitude: o.pickupAddress.longitude,
        });
      }
      if (dropStatuses.includes(o.status)) {
        candidates.push({
          orderId: o.id,
          orderNumber: o.orderNumber,
          type: 'dropoff',
          line1: o.dropoffAddress.line1,
          city: o.dropoffAddress.city,
          latitude: o.dropoffAddress.latitude,
          longitude: o.dropoffAddress.longitude,
        });
      }
    }

    const withCoords = candidates.filter((s) => s.latitude != null && s.longitude != null);
    const withoutCoords = candidates.filter((s) => s.latitude == null || s.longitude == null);

    let originLat = driver.currentLat;
    let originLng = driver.currentLng;
    if ((originLat == null || originLng == null) && withCoords.length > 0) {
      originLat = withCoords[0].latitude!;
      originLng = withCoords[0].longitude!;
    }

    const ordered: Stop[] = [];
    const remaining = [...withCoords];
    let curLat = originLat;
    let curLng = originLng;
    let totalKm = 0;

    while (remaining.length > 0 && curLat != null && curLng != null) {
      let bestIdx = 0;
      let bestDist = Infinity;
      for (let i = 0; i < remaining.length; i++) {
        const s = remaining[i];
        const dist = haversineKm(
          { lat: curLat, lng: curLng },
          { lat: s.latitude!, lng: s.longitude! },
        );
        if (dist < bestDist) {
          bestDist = dist;
          bestIdx = i;
        }
      }
      const next = remaining.splice(bestIdx, 1)[0];
      totalKm += bestDist === Infinity ? 0 : bestDist;
      ordered.push(next);
      curLat = next.latitude;
      curLng = next.longitude;
    }

    const stops = [...ordered, ...withoutCoords];

    return {
      stops,
      totalStops: stops.length,
      estimatedKm: Math.round(totalKm * 100) / 100,
    };
  }
}

export const driverOpsService = new DriverOpsService();
